### PR TITLE
Part2: Openiddict: Init OpenIddict & Database Migration & Auth Policies

### PR DESCRIPTION
**Pull request by Kukks**

### CODE REVIEW

File: BTCPayServer/Extensions/OpenIddictExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public static class OpenIddictExtensions
+    {
+        public static OpenIddictServerBuilder ConfigureSigningKey(this OpenIddictServerBuilder builder,
+            IConfiguration configuration)
+        {
+            
+            var file = Path.Combine(configuration.GetDataDir(), "rsaparams");
+                    
+            RSACryptoServiceProvider RSA = new RSACryptoServiceProvider(2048);
```
Comment (NicolasDorier) [inline BTCPayServer/Extensions/OpenIddictExtensions.cs]: Does this generate a key? If yes, we should not do this RSA key generation is super computationally expensive.

File: BTCPayServer/BTCPayServer.csproj [hunk 2]
```diff
@@ -58,6 +58,7 @@
     <PackageReference Include="NicolasDorier.StandardConfiguration" Version="1.0.0.18" />
     <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="2.1.2" />
     <PackageReference Include="OpenIddict" Version="2.0.0" />
+    <PackageReference Include="OpenIddict" Version="2.0.0" />
```
Comment (NicolasDorier) [inline BTCPayServer/BTCPayServer.csproj]: dups

File: BTCPayServer/Security/Policies.cs [hunk 3]
```diff
@@ -8,7 +8,9 @@ namespace BTCPayServer.Security
 {
     public static class Policies
     {
+        
         public const string BitpayAuthentication = "Bitpay.Auth";
+        public const string Dynamic = "dynamic";
```
Comment (NicolasDorier) [inline BTCPayServer/Security/Policies.cs]: Why dynamic? `OpenIddictAuthentication` would make better sense?

Comment (NicolasDorier) [reply BTCPayServer/Security/Policies.cs]: nevermind, I understand...

**Comment by NicolasDorier** [comment]
Can you rebase? I added some tests testing some authentication stuff.

File: BTCPayServer/Views/Account/Login.cshtml [hunk 4]
```diff
@@ -59,7 +60,7 @@
                     <h4>Use another service to log in.</h4>
                     <hr />
                     @{
-                        var loginProviders = (await SignInManager.GetExternalAuthenticationSchemesAsync()).ToList();
```
Comment (NicolasDorier) [inline BTCPayServer/Views/Account/Login.cshtml]: maybe we can cleanup all of this, we don't use it.

File: BTCPayServer/Extensions/OpenIddictExtensions.cs (hunk 1 above)
Comment (Kukks) [reply BTCPayServer/Extensions/OpenIddictExtensions.cs]: It is a one-time thing and only creates it if the file does not exist.

**Comment by Kukks** [comment]
rebased

File: BTCPayServer/Views/Account/Login.cshtml [hunk 5]
```diff
@@ -1,5 +1,6 @@
 ﻿@using System.Collections.Generic
 @using System.Linq
+@using BTCPayServer.Security
```
Comment (NicolasDorier) [inline BTCPayServer/Views/Account/Login.cshtml]: can probably remove this.

**Comment by Kukks** [comment]
updated @NicolasDorier

File: BTCPayServer/Hosting/BTCPayServerServices.cs [hunk 6]
```diff
@@ -245,9 +241,40 @@ public static IServiceCollection AddBTCPayServer(this IServiceCollection service
         
         private static void AddBtcPayServerAuthenticationSchemes(this IServiceCollection services, IConfiguration configuration)
         {
-            services.AddAuthentication()
+            JwtSecurityTokenHandler.DefaultInboundClaimTypeMap.Clear();
+            JwtSecurityTokenHandler.DefaultOutboundClaimTypeMap.Clear();
+
+            services.AddAuthentication(options =>
+                {
+                    options.DefaultAuthenticateScheme =  Policies.Dynamic;
```
Comment (NicolasDorier) [inline BTCPayServer/Hosting/BTCPayServerServices.cs]: I am not comfortable with this.
What if OpenIddict is not well configured, the hacker inject a `Authorization` header to force `JwtBearerDefaults.AuthenticationScheme`  authentication and gain access to everything?

Why can't we just precise explicitly at controller level, the policy we want to use?

Comment (Kukks) [reply BTCPayServer/Hosting/BTCPayServerServices.cs]: I've removed the Dynamic policy

**Pull request merged**